---
title: "scRNAseq_ETP_DN2"
author: "Ekaterina Takmakova"
date: "5 5 2021"
output: html_document
---


```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

# Setup the Seurat Object

```{r init}
library(dplyr)
library(Seurat)
library(sctransform)
library(patchwork)
library(ggplot2)
library(xlsx)
library(openxlsx)

# Load the dataset
thymus.data <- Read10X("C:/Users/koniaeva/R/seurat_GSE130812_ETP_DN2/data/rep_1")

# Initialize the Seurat object with the raw (non-normalized data).
thymus <- CreateSeuratObject(counts = thymus.data, project = "ETP_DN3", min.cells = 3, min.features = 200)
thymus
```

# QC and normalization of data

```{r, fig.height=4, fig.width=8}

# A list of cell cycle markers from Tirosh et al 2015
m.s.genes <- read.table("C:/Users/koniaeva/R/seurat_GSE130812_ETP_DN2/data/s.genes.murine.txt")
m.s.genes <- as.character(m.s.genes)
m.g2m.genes <- read.table("C:/Users/koniaeva/R/seurat_GSE130812_ETP_DN2/data/g2m.genes.murine.txt")
m.g2m.genes <- as.character(m.g2m.genes)

thymus <- CellCycleScoring(thymus, s.features = m.s.genes, g2m.features = m.g2m.genes)

# Mitochondrial genes migth are marked as 'mt', 'mt-' for mouse
thymus[["percent.mt"]] <- PercentageFeatureSet(thymus, pattern = "^mt")

# Visualize QC metrics as a violin plot.
# nFeature_RNA is the number of genes detected in each cell
# nCount_RNA is the total number of molecules detected within a cell
thymus[["log2.nUMI.1"]] <- log2(thymus[["nCount_RNA"]]+1)
VlnPlot(thymus, features = c("nFeature_RNA", "log2.nUMI.1", "percent.mt"), ncol = 3)

# Selection of data
thymus <- subset(thymus, subset = nFeature_RNA > 1000 & nFeature_RNA < 4000 & nCount_RNA < 1000000  & percent.mt < 6)

# SCTassay - Normalization of data
thymus <- SCTransform(thymus, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt"), return.only.var.genes = FALSE)

head(thymus[[]])
```

# Perform linear dimensional reduction (PCA)

```{r, fig.height=5, fig.width=12}

thymus <- RunPCA(thymus, verbose = FALSE)

# Examine and visualize PCA results a few different ways
print(thymus[['pca']], dims = 1:10, nfeatures = 10)
DimHeatmap(thymus, dims = 1:3, cells = 500, balanced = TRUE) + theme(text = element_text(size = 90))
VizDimLoadings(thymus, dims = 1:3, reduction = "pca", ncol = 3)

# Determine the 'dimensionality' of the dataset
ElbowPlot(thymus, ndims = 50)

```

# Cluster the cells, run non-linear dimensional reduction (UMAP)

```{r, fig.height=5, fig.width=6}

DefaultAssay(thymus) <- "SCT"
thymus <- RunUMAP(thymus, dims = 1:20)
thymus <- FindNeighbors(thymus, dims = 1:10)
thymus <- FindClusters(thymus, resolution = 0.9)

DimPlot(thymus, reduction = 'umap', label = FALSE, pt.size = 0.3, label.size = 5)

```

# Visualize specific markers to identify clusters as cell populations

```{r, fig.height=6, fig.width=15}

FeaturePlot(thymus, features = c("Kit", "Flt3", "Il2ra", "Bcl11b", "Il7r", "Rag1"), pt.size = 0.005, ncol = 4)
FeaturePlot(thymus, features = c("Kit", "Flt3", "Cd44", "Il2ra", "Lmo2", "Tcf7", "Gata3", "Bcl11b", "Il7r", "Cd3e", "Rag1", "Elane"), ncol = 4)
FeaturePlot(thymus, features = c("Lmo2", "Mef2c", "Cd34", "Cd7", "Cd52", "Hbb-bt", "Car2", "Evl", "Ly6d", "Thy1", "Irf7", "Lef1"), ncol = 4)
FeaturePlot(thymus, features = c("Ets1", "Notch3", "Rag1", "Ptcra", "lef1", "Cd3g", "Cd3d", "Cd3e", "Cd28", "Notch2", "Myl10", "Irf7", "Cbx3"), ncol = 4)

# Re-order clusters
thymus@active.ident <- factor(x = thymus@active.ident, levels = c("12", "4", "7", "6", "11", "3", "0", "5", "1", "9", "2", "8", "10"))

DotPlot(thymus, features = c("Id3", "Ptcra", "Rag1", "Ets1", "Thy1", "Il7r", "Bcl11b", "Il2ra", "Tcf7", "Flt3", "Kit", "Elane", "Notch1"), cols = c("grey", "blue"), col.min = -1, col.max = 1.5) + RotatedAxis() + coord_flip()

```

# Cluster annotation

```{r fig.height=5, fig.width=6}

thymus <- RenameIdents(thymus, "0" = "DN1", "1" = "DN2a", "2" = "DN2b", "3" = "DN1", "4" = "ETP", "5" = "DN2a", "6" = "ETP", "7" = "ETP", "8" = "DN3a", "9" = "DN2b", "10" = "DN3b", "11" = "DN1", "12" = "GrP")

thymus@active.ident <- factor(x = thymus@active.ident, levels = c("ETP", "DN1", "DN2a", "DN2b", "DN3a", "DN3b", "GrP"))
thymus <- subset(thymus, idents = c("ETP", "DN1", "DN2a", "DN2b", "DN3a", "DN3b"))

DimPlot(thymus, reduction = 'umap', label = FALSE, pt.size = 0.3, label.size = 5)

```

# Finding differentially expressed features (cluster biomarkers)

```{r fig.height=13, fig.width=10}

thymus.markers <- FindAllMarkers(thymus, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
thymus.markers %>% group_by(cluster) %>% slice_max(n = 10, order_by = avg_log2FC) -> top10

top10
DoHeatmap(thymus, features = top10$gene) + theme(text = element_text(size = 20))

```

# Visualize specific markers

```{r fig.height=3, fig.width=12}

VlnPlot(thymus, features = c("Flt3", "Erg", "Runx2", "Tcf7l2", "Lmo2", "Hoxa9", "Mef2c", "Cd34", "Bcl11a", "Spi1", "Kit", "Mycn", "Lyl1", "Cebpb", "Hhex", "Notch1", "Hes1"), stack = TRUE, flip = TRUE) + NoLegend()

VlnPlot(thymus, features = c("Tcf7", "Runx1", "Il2ra", "Gata3", "Ly6d", "Il7r", "Bcl11b", "Ets1", "Thy1", "Notch3", "Rag1", "Cd3g", "Cd3d", "Cd3e", "Ptcra", "Lef1", "Id3"), stack = TRUE, flip = TRUE) + NoLegend()

DotPlot(thymus, features = c("Id3", "Lef1", "Ptcra", "Cd3e", "Cd3d", "Cd3g", "Rag1", "Notch3", "Thy1", "Ets1", "Bcl11b", "Il7r", "Ly6d", "Gata3", "Il2ra", "Runx1", "Tcf7", "Hes1", "Notch1", "Hhex", "Cebpb", "Lyl1", "Mycn", "Kit", "Spi1", "Bcl11a", "Cd34", "Mef2c", "Hoxa9", "Lmo2", "Tcf7l2", "Runx2", "Erg", "Flt3")) + RotatedAxis() + coord_flip()

DotPlot(thymus, features = c("Il4ra", "Il12rb2", "Il12rb1", "Il18rap", "Il18r1", "Igf2r", "Igf1r", "Kit", "Il7r", "Flt3")) + RotatedAxis() + coord_flip()

DotPlot(thymus, features = c("Nrip1", "Parp8", "Hnrnpul1", "Tgfb1", "Ccdc97", "Axl")) + RotatedAxis() + coord_flip()

AverageExpression(thymus, assays = "SCT", slot = "data", features = c("Il7r", "Nrip1", "Parp8", "Hnrnpul1", "Tgfb1", "Ccdc97", "Axl"))

```
